### PR TITLE
Hash based lookup implementation

### PR DESCRIPTION
**Pull request by guliyevemil1**

### CODE REVIEW

**Comment by guliyevemil1** [comment]
If you want me to add comments to explain the logic a bit, let me know. I basically took what you were doing with regexes and just did it via string manipulation and this made it much faster.

**Comment by guliyevemil1** [comment]
Btw, there was a small bug that's kind of a gotcha in Go if you haven't seen it before. The logic for figuring out `bestRule` was actually buggy because the temporary variable `rule` is reused throughout the for loop and overwritten on each iteration. You can see the buggy behavior demonstrated here:

https://play.golang.org/p/72PdGvRy9Dh

**Comment by guliyevemil1** [comment]
Hey, can I get a timeline of when you think you can merge this? I'd like to use your library but this is a blocker. Thanks!

**Comment by weppos** [comment]
I need to allocate some time to look into it. I will try to find some time within this week.

**Comment by guliyevemil1** [comment]
Hey man, I understand you're busy. I found your library from the official publicsuffix.org list and I liked your API better than x/net package. The performance was unacceptable for my use case so I decided to help you and improve it. It passes all your tests and the benchmarks are comparable or better than x/net (which I added in the first commit). Can you please spare some time soon to merge this? I would like to continue using your library instead of being forced to use my forked off version that I don't think I can maintain.

File: publicsuffix/publicsuffix.go [hunk 1]
```diff
@@ -195,43 +195,23 @@ Scanning:
...
-
-func (l *List) selectRules(name string, options *FindOptions) []Rule {
-	var found []Rule
-
-	// In this phase the search is a simple sequential scan
-	for _, rule := range l.rules {
-		if !rule.Match(name) {
-			continue
+	for {
+		rule, ok := l.rules[name]
```
Comment (weppos) [inline publicsuffix/publicsuffix.go]: There is a subtle but important implication in this change, and I want to add a note here mostly as a reminder.
This change will assume that `*.foo` and `foo` won't exist in the same definition.

Please see the main ticket for further information.

**Comment by weppos** [comment]
There is a subtle but important implication in this change, and I want to add a note here mostly as a reminder.

This change will assume that `*.foo` and `foo` won't exist in the same definition. This assumption may not leave forever, and may actually not even be correct.
https://groups.google.com/d/topic/publicsuffix-discuss/rY6yRK802Kg/discussion

For better or for worse, this is the same current limitation of
https://github.com/weppos/publicsuffix-ruby

Depending on the outcome of the discussion, the index key may have to be reshaped.
I actually also considered the idea of storing rules in a tree rather than a hash https://github.com/weppos/publicsuffix-ruby/pull/134, that could be a potential solution as finding the proper rule would be walking a tree, and `com/foo/*` could be different than `com/foo`.

File: publicsuffix/rules.go [hunk 2]
```diff
@@ -8626,5 +8626,7 @@ func init() {
 		{1, "now.sh", 2, true},
 		{1, "zone.id", 2, true},
 	}
-	DefaultList.rules = r[:]
+	for i := range r {
+		DefaultList.AddRule(&r[i])
```
Comment (weppos) [inline publicsuffix/rules.go]: What's the reason of this change? Is it because of #141?

Comment (weppos) [reply publicsuffix/rules.go]: OK, nevermind. I actually figured it out, we need to pass via `AddRule` because the internal `rule` is no longer a list.

My bad, sorry for the stupid question.

I need to determine the cost of this operation. It could be possible that it's more efficient to compile directly the `r` in the appropriate form and then assign.

Comment (guliyevemil1) [reply publicsuffix/rules.go]: I can't speak with absolute certainty, but my gut feeling is that this implementation eases the load on the garbage collector because nothing in this list should ever be garbage collected which means it's better to have all of them in a single giant array rather than allocating memory for each individually on the heap. I might be wrong and it might be worth benchmarking.

**Comment by guliyevemil1** [comment]
Thank you so much for taking the time to review the change and merging it!
